File: 501. Find Mode in Binary Search Tree/C++/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> findMode(TreeNode *root) {
        class Utils {
        public:
            void mode(TreeNode *root, vector<int> &result, int &prev, int &currentCount, int &maxCount) {
                if (!root) {
                    return;
                }
                mode(root->left, result, prev, currentCount, maxCount);
                if (prev == root->val) {
                    currentCount++;
                } else {
                    currentCount = 1;
                }
                prev = root->val;
                if (currentCount > maxCount) {
                    maxCount = currentCount;
                    result.clear();
                    result.push_back(root->val);
                } else if (currentCount == maxCount) {
                    result.push_back(root->val);
                }
                mode(root->right, result, prev, currentCount, maxCount);
            }
        };
        Utils utils;
        vector<int> result;
        int prev;
        int currentCount = 0;
        int maxCount = 0;
        utils.mode(root, result, prev, currentCount, maxCount);
        return result;
    }
};

File: 501. Find Mode in Binary Search Tree/Java/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int[] findMode(TreeNode root) {
        class Utils {
            Integer prev = null;
            int currentCount = 0;
            int maxCount = 0;

            public void mode(TreeNode root, List<Integer> resultList) {
                if (root == null) {
                    return;
                }
                mode(root.left, resultList);
                if (prev != null && prev == root.val) {
                    currentCount++;
                } else {
                    currentCount = 1;
                }
                prev = root.val;
                if (currentCount > maxCount) {
                    maxCount = currentCount;
                    resultList.clear();
                    resultList.add(root.val);
                } else if (currentCount == maxCount) {
                    resultList.add(root.val);
                }
                mode(root.right, resultList);
            }
        }
        Utils utils = new Utils();
        List<Integer> resultList = new ArrayList<>();
        utils.mode(root, resultList);
        int[] result = new int[resultList.size()];
        for (int i = 0; i < resultList.size(); i++) {
            result[i] = resultList.get(i);
        }
        return result;
    }
}